File: string/Solution12.java
class Solution12 {
    public String intToRoman(int num) {
    	int[] digit1 = new int[]{1, 4, 5, 9, 10, 40, 50, 90, 100, 400, 500, 900, 1000};
    	String[] digit2 = new String[]{"I", "IV", "V", "IX", "X", "XL", "L", "XC", "C", "CD", "D", "CM", "M"};
    	String result = "";
   loop: while(num > 0){
    		for(int i = 0; i < digit1.length; i++){
    			if(i == digit1.length - 1 || ( num >= digit1[i] && num < digit1[i + 1])){
    				result += digit2[i];
    				num = num - digit1[i];
    				continue loop;
    			}
    		}
    	}
    	return result;
    }

    public static void main(String[] args){
    	Solution12 s = new Solution12();
    	System.out.println(s.intToRoman(9));
        System.out.println(s.intToRoman(58));
        System.out.println(s.intToRoman(20));
        System.out.println(s.intToRoman(1994));
    }
}

File: C++/Solution12.cpp
#include<iostream>
#include<vector>


using namespace std;

class Solution12 {
public:
    string intToRoman(int num) {
        vector<pair<int, string>> map = {
            {1, "I"}, {4, "IV"}, {5, "V"}, {9, "IX"}, {10, "X"}, {40, "XL"}, {50, "L"}, {90, "XC"}, 
            {100, "C"}, {400, "CD"}, {500, "D"}, {900, "CM"}, {1000, "M"}
        };
        string result = "";
        int size = map.size();
        while(num > 0){
            for(int i = size - 1; i >= 0; i--){
                auto cur = map[i];
                if(num >= cur.first){
                    num = num - cur.first;
                    result += cur.second;
                    break;
                }
            }
        }
        return result;
    }
};

int main(){
    Solution12 s;
    cout << s.intToRoman(1994) << endl;
    return 0;
}